scipy/spatial/src/distance_metrics.h

```diff
@@ -345,3 +345,431 @@ struct CanberraDistance {
         });
     }
 };
+
+struct HammingDistance {
+    template <typename T>
+    struct Acc {
+        Acc(): nonmatches(0), total(0) {}
+        T nonmatches, total;
+    };
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        transform_reduce_2d_<4>(out, x, y, [](T x, T y) INLINE_LAMBDA {
+            Acc<T> acc;
+            acc.nonmatches = x != y;
+            acc.total = 1;
+            return acc;
+        },
+        [](const Acc<T>& acc) INLINE_LAMBDA {
+            return acc.nonmatches / acc.total;
+        },
+        [](const Acc<T>& a, const Acc<T>& b) INLINE_LAMBDA {
+            Acc<T> acc;
+            acc.nonmatches = a.nonmatches + b.nonmatches;
+            acc.total = a.total + b.total;
+            return acc;
+        });
+    }
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y, StridedView2D<const T> w) const {
+        transform_reduce_2d_(out, x, y, w, [](T x, T y, T w) INLINE_LAMBDA {
+            Acc<T> acc;
+            acc.nonmatches = w * (x != y);
+            acc.total = w;
+            return acc;
+        },
+        [](const Acc<T>& acc) INLINE_LAMBDA {
+            return acc.nonmatches / acc.total;
+        },
+        [](const Acc<T>& a, const Acc<T>& b) INLINE_LAMBDA {
+            Acc<T> acc;
+            acc.nonmatches = a.nonmatches + b.nonmatches;
+            acc.total = a.total + b.total;
+            return acc;
+        });
+    }
+};
+
+struct DiceDistance {
+    template <typename T>
+    struct Acc {
+        Acc(): nonmatches(0), tt_matches(0) {}
+        T nonmatches, tt_matches;
+    };
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        transform_reduce_2d_<2>(out, x, y, [](T x, T y) INLINE_LAMBDA {
+            Acc<T> acc;
+            acc.nonmatches = (x != y);
+            acc.tt_matches = (x != 0) & (y != 0);
```

Comment:
These functions need to be changed slightly to deal with floating point inputs correctly. Basically you need to write everything in terms of logical and which becomes multiplication, and logical not which becomes `1.0 - x`. 
```suggestion
            acc.nonmatches = x * (1.0 - y) + y * (1.0 - x);
            acc.tt_matches = x * y;
```

There is an existing bug in cdist that probably means it isn't tested properly. See #17703

Reply:
I rebased, added a regression test for that issue, and the suggested patch that allows the test to pass, and force pushed in. Maybe better now?